yktrace/src/tir.rs

```diff
@@ -285,74 +258,143 @@ impl TirTrace {
+    }
+
+    fn rename_rvalue(&self, rvalue: &Rvalue) -> Rvalue {
         match rvalue {
             Rvalue::Use(op) => {
-                let newop = TirTrace::rename_operand(rename_map, op);
+                let newop = self.rename_operand(op);
                 Rvalue::Use(newop)
             }
             Rvalue::BinaryOp(binop, op1, op2) => {
-                let newop1 = TirTrace::rename_operand(rename_map, op1);
-                let newop2 = TirTrace::rename_operand(rename_map, op2);
+                let newop1 = self.rename_operand(op1);
+                let newop2 = self.rename_operand(op2);
                 Rvalue::BinaryOp(binop.clone(), newop1, newop2)
             }
             Rvalue::CheckedBinaryOp(binop, op1, op2) => {
-                let newop1 = TirTrace::rename_operand(rename_map, op1);
-                let newop2 = TirTrace::rename_operand(rename_map, op2);
+                let newop1 = self.rename_operand(op1);
+                let newop2 = self.rename_operand(op2);
                 Rvalue::CheckedBinaryOp(binop.clone(), newop1, newop2)
             }
             Rvalue::Unimplemented(_) => rvalue.clone()
         }
     }
-    fn rename_operand(rename_map: &HashMap<Local, Vec<Local>>, operand: &Operand) -> Operand {
+
+    fn rename_operand(&self, operand: &Operand) -> Operand {
         match operand {
-            Operand::Place(p) => Operand::Place(TirTrace::rename_place(rename_map, p)),
+            Operand::Place(p) => Operand::Place(self.rename_place(p)),
             Operand::Constant(_) => operand.clone()
         }
     }
 
-    fn rename_place(rename_map: &HashMap<Local, Vec<Local>>, place: &Place) -> Place {
-        // In the future there should always be a mapping for any local in the trace.
-        // Unfortunately, since we are still getting remnants of the trace header and tail in our
-        // trace, this is not always the case. So for now print warnings and don't attempt to
-        // rename those locals.
-        match rename_map.get(&place.local) {
-            Some(v) => match v.last() {
-                Some(l) => {
-                    let mut p = place.clone();
-                    p.local = l.clone();
-                    p
-                }
-                None => {
-                    eprintln!("warning: mapping for {} already empty", &place.local);
-                    place.clone()
-                }
-            },
-            None => {
-                if &place.local != &Local(0) {
-                    // Local(0) is used for function returns and is thus never
-                    // renamed, so we don't need to print a warning for it.
-                    eprintln!("warning: could not find mapping for {}", &place.local);
-                }
-                place.clone()
-            }
+    fn rename_place(&self, place: &Place) -> Place {
+        if &place.local == &Local(0) {
+            // Local(0) is used for returning values from calls, so it doesn't need to be renamed.
```

Comment:
This feels fishy.

Suppose we have:
```
fn f() {
  $0 = 5;
  $1 = g(); // $1 unused
}

fn g() {
  $0 = 6;
}
```

Wouldn't `g`'s $0 obliterate `f`'s?

Reply:
To the best of my knowledge, `$0` is only ever used to return variables from a function so it's always the last instruction of a block. A function `f` as above would thus not be possible.

Reply:
1. For MIR after deaggregation there can be multiple statements each assigning to a different part of `_0`.
2. Also for MIR there is https://github.com/rust-lang/rust/pull/72205 that will hopefully be merged soon(ish).

Reply:
>     1. For MIR after deaggregation there can be multiple statements each assigning to a different part of `_0`.

This should be fine as long as the `_0` is the return variable and not used for anything else.

>     2. Also for MIR there is [rust-lang/rust#72205](https://github.com/rust-lang/rust/pull/72205) that will hopefully be merged soon(ish).

Does this mean that as of now `_0` can be used as a normal variable, but this PR is going to fix that?

Reply:
>[$0 is] always the last instruction of a block.

I'm not sure if that's true or not.

Reply:
> Does this mean that as of now _0 can be used as a normal variable, but this PR is going to fix that?

Rather the opposite. Before the PR the return value was copied to `_0` just before returning. After the PR the return value may be written to `_0` and then be read from within the same function at a later point.

```rust
bb0: {
    _1 = const 0u8;
    _2 = Add(_1, const 0u8);
    SwitchInt(_2) -> [0: bb1, otherwise: bb2];
}
bb1: {
    _0 = _1;
    return;
}
bb2: {
    goto -> bb2;
}
```

might turn into

```rust
bb0: {
    _0 = const 0u8;
    _2 = Add(_0, const 0u8);
    SwitchInt(_2) -> [0: bb1, otherwise: bb2];
}
bb1: {
    return;
}
bb2: {
    goto -> bb2;
}
```

if I understand it correctly.

Reply:
Ah, thanks for the clarification. That's good to know as this will mess with inlined functions. But I think I'll do that in a separate PR and in the meantime add an `assert` to check that `_0` is not overwritten by an inlined function.